File: DSA/DSA_450/03_string/035_Min_Chars_Front_Palindrome.cpp
/*
Problem: Minimum Characters to Add at Front to Make String Palindrome
URL: https://www.geeksforgeeks.org/minimum-characters-added-front-make-string-palindrome/

Problem Statement:
Given a string str, find the minimum number of characters to be added at the
front to make the string a palindrome.

Sample Input/Output:
Input: "AACECAAAA"
Output: 2

Input: "ABC"
Output: 2
*/

#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int Min_Chars_LPS(string str) {
        /*
        Using KMP LPS array on str + "$" + reverse(str)
        Time Complexity: O(n)
        Space Complexity: O(n)
        */
        string rev = str;
        reverse(rev.begin(), rev.end());
        string concat = str + "$" + rev;
        int n = concat.size();

        vector<int> lps(n, 0);
        int len = 0, i = 1;
        while (i < n) {
            if (concat[i] == concat[len]) {
                len++;
                lps[i] = len;
                i++;
            } else {
                if (len != 0) len = lps[len - 1];
                else { lps[i] = 0; i++; }
            }
        }
        return str.size() - lps[n - 1];
    }

    int Min_Chars_Brute(string str) {
        /*
        Keep removing last char until palindrome
        Time Complexity: O(n^2)
        Space Complexity: O(n)
        */
        int cnt = 0;
        string s = str;
        while (!s.empty()) {
            string rev = s;
            reverse(rev.begin(), rev.end());
            if (s == rev) break;
            s.pop_back();
            cnt++;
        }
        return cnt;
    }

    int Min_Chars_Two_Pointer(string str) {
        /*
        Two pointer - find longest palindromic prefix
        Time Complexity: O(n^2) worst case
        Space Complexity: O(n)
        */
        int n = str.size();
        int i = 0, j = n - 1;
        int suffixEnd = n - 1;

        while (i < j) {
            if (str[i] == str[j]) {
                i++;
                j--;
            } else {
                i = 0;
                suffixEnd--;
                j = suffixEnd;
            }
        }
        return n - suffixEnd - 1;
    }
};

void Test_Min_Chars_Front_Palindrome() {
    Solution sol;
    vector<string> tests = {"AACECAAAA", "ABC", "BABABAA", "a", "ab", "aaa"};

    for (auto& s : tests) {
        cout << "Input: " << s << endl;
        cout << "LPS: " << sol.Min_Chars_LPS(s) << endl;
        cout << "Brute: " << sol.Min_Chars_Brute(s) << endl;
        cout << "Two Pointer: " << sol.Min_Chars_Two_Pointer(s) << endl;
        cout << string(50, '-') << endl;
    }
}

int main() {
    Test_Min_Chars_Front_Palindrome();
    return 0;
}

File: DSA/DSA_450/Python/03_string/035_Min_Chars_Front_Palindrome.py
"""
Problem: Minimum Characters to Add at Front to Make String Palindrome
URL: https://www.geeksforgeeks.org/minimum-characters-added-front-make-string-palindrome/

Problem Statement:
Given a string str, find the minimum number of characters to be added at the
front to make the string a palindrome.

Sample Input/Output:
Input: "AACECAAAA"
Output: 2

Input: "ABC"
Output: 2
"""


class Solution:
    def Min_Chars_LPS(self, str_val):
        """
        Using KMP LPS array on str + "$" + reverse(str)
        Time Complexity: O(n)
        Space Complexity: O(n)
        """
        rev = str_val[::-1]
        concat = str_val + "$" + rev
        n = len(concat)

        lps = [0] * n
        length = 0
        i = 1
        while i < n:
            if concat[i] == concat[length]:
                length += 1
                lps[i] = length
                i += 1
            else:
                if length != 0:
                    length = lps[length - 1]
                else:
                    lps[i] = 0
                    i += 1

        return len(str_val) - lps[n - 1]

    def Min_Chars_Brute(self, str_val):
        """
        Keep removing last char until palindrome
        Time Complexity: O(n^2)
        Space Complexity: O(n)
        """
        cnt = 0
        s = str_val
        while s:
            rev = s[::-1]
            if s == rev:
                break
            s = s[:-1]
            cnt += 1
        return cnt

    def Min_Chars_Two_Pointer(self, str_val):
        """
        Two pointer - find longest palindromic prefix
        Time Complexity: O(n^2) worst case
        Space Complexity: O(n)
        """
        n = len(str_val)
        i = 0
        j = n - 1
        suffixEnd = n - 1

        while i < j:
            if str_val[i] == str_val[j]:
                i += 1
                j -= 1
            else:
                i = 0
                suffixEnd -= 1
                j = suffixEnd

        return n - suffixEnd - 1


def Test_Min_Chars_Front_Palindrome():
    sol = Solution()
    tests = ["AACECAAAA", "ABC", "BABABAA", "a", "ab", "aaa"]

    for s in tests:
        print(f"Input: {s}")
        print(f"LPS: {sol.Min_Chars_LPS(s)}")
        print(f"Brute: {sol.Min_Chars_Brute(s)}")
        print(f"Two Pointer: {sol.Min_Chars_Two_Pointer(s)}")
        print('-' * 50)


if __name__ == "__main__":
    Test_Min_Chars_Front_Palindrome()
